### ISSUE TITLE
Engine writes past the end of the visible entities list used by player movement code

### ISSUE THREAD

**Issue by SamVanheer** [ValveSoftware/halflife#3349]
When the engine populates the `playermove_t` struct prior to executing the player movement code it copies entity data to a couple lists here:
https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/pm_shared/pm_defs.h#L162-L171

The engine checks to make sure it doesn't write past the end of the `physents` and `moveents` arrays but it doesn't check this for the `visents` array.

If there are a lot of visible entities then it will overwrite the variables right behind it:
https://github.com/ValveSoftware/halflife/blob/c7240b965743a53a29491dd49320c88eecf6257b/pm_shared/pm_defs.h#L173-L220

Some of those are data, some are function pointers most of which will point to an invalid address. The game crashes when the data is used or when an invalid function pointer is accessed.

One case where this happens is when a `monstermaker` set to spawn infinite monsters immediately intersects with a moving `func_door` that deals enough damage to instantly gib the spawned monster. Newly spawned monsters are instantly gibbed, spawning several gib entities.

These will quickly fill up the `visents` array on the server side. The server only sends up 255 visible entities so the client is far less likely to encounter this issue.

Fixing this requires the engine to stop adding entities to the `visents` array if it's full.

Note that the array's size caps out at 600 entities. The default entity limit is 900 (more in multiplayer) and the engine supports up to 2048 visible entities before the networking code causes a client-side crash to occur. 

The physics code won't work properly once the 600 entity limit has been exceeded since it can't "see" the additional entities, but since users haven't reported this kind of crash before it's unlikely to cause any problems. The game will simply continue to work instead of crashing, albeit without being able to deal with the extra entities.

**Comment by teddiboy** [comment]
that's right, when I create a lot of entities on the map, they are invisible, will be nice if correct you this :D
